subsys/bluetooth/services/dis.c

```diff
@@ -45,105 +48,265 @@ static struct dis_pnp dis_pnp_id = {
 			       read_pnp_id, NULL, &dis_pnp_id),
 #endif
 
 #if defined(CONFIG_BT_DIS_SERIAL_NUMBER)
-	BT_GATT_CHARACTERISTIC(BT_UUID_DIS_SERIAL_NUMBER,
-			       BT_GATT_CHRC_READ, BT_GATT_PERM_READ,
-			       read_str, NULL,
-			       BT_DIS_SERIAL_NUMBER_STR_REF),
+	BT_GATT_CHARACTERISTIC(BT_UUID_DIS_SERIAL_NUMBER, BT_GATT_CHRC_READ, BT_GATT_PERM_READ,
+			       read_str, NULL, BT_DIS_SERIAL_NUMBER_STR_REF),
 #endif
 #if defined(CONFIG_BT_DIS_FW_REV)
-	BT_GATT_CHARACTERISTIC(BT_UUID_DIS_FIRMWARE_REVISION,
-			       BT_GATT_CHRC_READ, BT_GATT_PERM_READ,
+	BT_GATT_CHARACTERISTIC(BT_UUID_DIS_FIRMWARE_REVISION, BT_GATT_CHRC_READ, BT_GATT_PERM_READ,
 			       read_str, NULL, BT_DIS_FW_REV_STR_REF),
 #endif
 #if defined(CONFIG_BT_DIS_HW_REV)
-	BT_GATT_CHARACTERISTIC(BT_UUID_DIS_HARDWARE_REVISION,
-			       BT_GATT_CHRC_READ, BT_GATT_PERM_READ,
+	BT_GATT_CHARACTERISTIC(BT_UUID_DIS_HARDWARE_REVISION, BT_GATT_CHRC_READ, BT_GATT_PERM_READ,
 			       read_str, NULL, BT_DIS_HW_REV_STR_REF),
 #endif
 #if defined(CONFIG_BT_DIS_SW_REV)
-	BT_GATT_CHARACTERISTIC(BT_UUID_DIS_SOFTWARE_REVISION,
-			       BT_GATT_CHRC_READ, BT_GATT_PERM_READ,
+	BT_GATT_CHARACTERISTIC(BT_UUID_DIS_SOFTWARE_REVISION, BT_GATT_CHRC_READ, BT_GATT_PERM_READ,
 			       read_str, NULL, BT_DIS_SW_REV_STR_REF),
 #endif
+#if defined(CONFIG_BT_DIS_UDI)
+	BT_GATT_CHARACTERISTIC(BT_UUID_UDI_FOR_MEDICAL_DEVICES, BT_GATT_CHRC_READ,
+			       BT_GATT_PERM_READ, read_udi, NULL, NULL),
+#endif
+#if defined(CONFIG_BT_DIS_SYSTEM_ID)
+	BT_GATT_CHARACTERISTIC(BT_UUID_DIS_SYSTEM_ID, BT_GATT_CHRC_READ, BT_GATT_PERM_READ,
+			       read_system_id, NULL, NULL),
+#endif
+#if defined(CONFIG_BT_DIS_IEEE_RCDL)
+	BT_GATT_CHARACTERISTIC(BT_UUID_GATT_IEEE_RCDL, BT_GATT_CHRC_READ, BT_GATT_PERM_READ,
+			       read_str, NULL, BT_DIS_IEEE_RCDL_STR_REF),
+#endif
 
 );
 
 #if defined(CONFIG_BT_DIS_SETTINGS)
-static int dis_set(const char *name, size_t len_rd,
-		   settings_read_cb read_cb, void *store)
+#if defined(CONFIG_BT_DIS_UDI)
+#define ADD_NULLTERM(x) (x > 0 ? x + sizeof('\0') : 0)
+static void dis_update_udi_value(const char *new, char *old, settings_read_cb read_cb,
+				 const char *logkey)
+{
+	/*
+	 * The characteristic contains 1 one-byte flag and 4-null-terminated.
+	 * The null-terminators are only present for strings that are in the flags.
+	 */
+	const size_t merged_size = sizeof(uint8_t) +
+				   ADD_NULLTERM(strlen(BT_DIS_UDI_LABEL_STR_REF)) +
+				   ADD_NULLTERM(strlen(BT_DIS_UDI_DI_STR_REF)) +
+				   ADD_NULLTERM(strlen(BT_DIS_UDI_ISSUER_STR_REF)) +
+				   ADD_NULLTERM(strlen(BT_DIS_UDI_AUTHORITY_STR_REF));
+
+	size_t without_old = merged_size - ADD_NULLTERM(strlen(old));
+
+	bool valid = BT_ATT_MAX_ATTRIBUTE_LEN >= without_old + ADD_NULLTERM(strlen(new));
+
+	if (!valid) {
+		LOG_ERR("Failed to set UDI %s. Not enough space. The sum of the 4 DIS "
+			"UDI for "
+			"Medical Devices strings may not exceed the maximum attribute "
+			"length.",
+			logkey);
```

Comment:
```suggestion
		LOG_ERR("Failed to set UDI %s. Not enough space. The sum of the 4 DIS UDI for "
			"Medical Devices strings may not exceed the maximum attribute length.",
			logkey);
```

Max line length in Zephyr is 100 characters